pkg/grpc_middleware/dialer/builder.go

```diff
@@ -0,0 +1,319 @@
+package dialer
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"log"
+	"net/url"
+	"strconv"
+	"time"
+
+	"github.com/caring/go-packages/v2/pkg/errors"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/backoff"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/keepalive"
+)
+
+const (
+	// MinConnectTimeout before attempting reconnect.
+	MinConnectTimeout = 20 * time.Second
+)
+
+type Builder struct {
+	options         []grpc.DialOption
+	enabledBlocking bool
+	connectParams   grpc.ConnectParams
+	keepAliveParams keepalive.ClientParameters
+	credentials     credentials.PerRPCCredentials
+	tlsConfig       *tls.Config
+	uinterceptors   []grpc.UnaryClientInterceptor
+	sinterceptors   []grpc.StreamClientInterceptor
+	dns             *string
+	port            *string
+	withTLS         *bool
+}
+
+// WithOptions allows possing in multiple grpc dial options
+// DialOption configures how we set up the connection.
+func (b *Builder) WithOptions(opts ...grpc.DialOption) {
```

Comment:
I do like the idea of making fields public, it goes more along the lines with our best practices everywhere else of not having `New` methods unless needed, and instead configuring by setting fields.

Reply:
The With fields seem to be a "nice to have" to decouple the package receiving a builder from go-packages.  
see https://github.com/caring/call-scoring/blob/jlundy/new-dial-builder/api/graphql/resolvers.go#L31-L36

though TBH other than here where i have two conflicting versions of go-packages included into that branch i'm not sure where we would run into the issue.